File: kernel/src/editor/editor_manager.cpp
#include "jarvis/editor/editor_manager.h"
#include <fstream>
#include <filesystem>
#include <unordered_map>
#include <algorithm>

namespace jarvis::editor {
namespace fs = std::filesystem;

static const std::unordered_map<std::string, std::string> LANG_MAP = {
    {".js", "javascript"}, {".jsx", "javascript"},
    {".ts", "typescript"}, {".tsx", "typescript"},
    {".mjs", "javascript"}, {".cjs", "javascript"},
    {".py", "python"},
    {".cpp", "cpp"}, {".cc", "cpp"}, {".cxx", "cpp"},
    {".hpp", "cpp"}, {".hh", "cpp"}, {".hxx", "cpp"},
    {".h", "c"},
    {".c", "c"},
    {".java", "java"},
    {".rs", "rust"},
    {".go", "go"},
    {".rb", "ruby"},
    {".php", "php"},
    {".swift", "swift"},
    {".kt", "kotlin"}, {".kts", "kotlin"},
    {".scala", "scala"},
    {".dart", "dart"},
    {".lua", "lua"},
    {".pl", "perl"}, {".pm", "perl"},
    {".r", "r"},
    {".json", "json"},
    {".md", "markdown"}, {".mdx", "markdown"},
    {".html", "html"}, {".htm", "html"},
    {".css", "css"}, {".scss", "scss"}, {".less", "less"},
    {".xml", "xml"}, {".svg", "xml"},
    {".yaml", "yaml"}, {".yml", "yaml"},
    {".toml", "toml"},
    {".sql", "sql"},
    {".sh", "shell"}, {".bash", "shell"}, {".zsh", "shell"},
    {".bat", "bat"}, {".cmd", "bat"},
    {".ps1", "powershell"},
    {".dockerfile", "dockerfile"},
    {".gitignore", "ignore"},
    {".env", "plaintext"},
    {".txt", "plaintext"},
    {".csv", "plaintext"},
    {".log", "plaintext"},
    {".diff", "diff"}, {".patch", "diff"},
};

static std::string detect(const std::string& filename) {
    auto pos = filename.find_last_of('.');
    if (pos == std::string::npos) return "plaintext";

    auto ext = filename.substr(pos);
    std::string lower;
    lower.reserve(ext.size());
    for (auto c : ext) lower += static_cast<char>(std::tolower(c));

    auto it = LANG_MAP.find(lower);
    if (it != LANG_MAP.end()) return it->second;

    auto nameLower = filename;
    std::transform(nameLower.begin(), nameLower.end(), nameLower.begin(), ::tolower);
    auto dit = LANG_MAP.find(nameLower);
    if (dit != LANG_MAP.end()) return dit->second;

    return "plaintext";
}

class EditorManager : public IEditorService {
public:
    std::optional<EditorTabInfo> openFile(const std::string& path) override {
        if (!fs::exists(path)) return std::nullopt;

        std::ifstream file(path, std::ios::binary | std::ios::ate);
        if (!file.is_open()) return std::nullopt;

        auto size = file.tellg();
        file.seekg(0);

        std::string content(static_cast<size_t>(size), '\0');
        file.read(content.data(), size);

        auto lastModified = fs::last_write_time(path).time_since_epoch().count();

        EditorTabInfo info;
        info.path = path;
        info.language = detect(path);
        info.content = content;
        info.size = static_cast<int64_t>(size);
        info.lastModified = static_cast<int64_t>(lastModified);
        info.isDirty = false;

        openTabs_[path] = info;
        return info;
    }

    bool saveFile(const std::string& path, const std::string& content) override {
        std::ofstream file(path, std::ios::binary | std::ios::trunc);
        if (!file.is_open()) return false;

        file.write(content.data(), static_cast<std::streamsize>(content.size()));
        if (!file.good()) return false;

        if (openTabs_.count(path)) {
            openTabs_[path].content = content;
            openTabs_[path].isDirty = false;
            openTabs_[path].size = static_cast<int64_t>(content.size());
            openTabs_[path].lastModified =
                static_cast<int64_t>(fs::last_write_time(path).time_since_epoch().count());
        }
        return true;
    }

    bool closeFile(const std::string& path) override {
        return openTabs_.erase(path) > 0;
    }

    std::vector<EditorTabInfo> getOpenFiles() override {
        std::vector<EditorTabInfo> result;
        result.reserve(openTabs_.size());
        for (const auto& [_, tab] : openTabs_) {
            result.push_back(tab);
        }
        return result;
    }

    std::string detectLanguage(const std::string& filename) override {
        return detect(filename);
    }

    void updateTabState(const std::string& path, bool isDirty) override {
        if (openTabs_.count(path)) {
            openTabs_[path].isDirty = isDirty;
        }
    }

private:
    std::unordered_map<std::string, EditorTabInfo> openTabs_;
};

IEditorService* createEditorService() {
    return new EditorManager();
}

} // namespace jarvis::editor

File: backend/jarvis/editor_manager.py
"""Editor service — file open/save/close, language detection, settings."""

import os
from dataclasses import dataclass

from jarvis.database import Database

_LANG_MAP: dict[str, str] = {
    ".js": "javascript", ".jsx": "javascript",
    ".ts": "typescript", ".tsx": "typescript",
    ".mjs": "javascript", ".cjs": "javascript",
    ".py": "python",
    ".cpp": "cpp", ".cc": "cpp", ".cxx": "cpp",
    ".hpp": "cpp", ".hh": "cpp", ".hxx": "cpp",
    ".h": "c",
    ".c": "c",
    ".java": "java",
    ".rs": "rust",
    ".go": "go",
    ".rb": "ruby",
    ".php": "php",
    ".swift": "swift",
    ".kt": "kotlin", ".kts": "kotlin",
    ".scala": "scala",
    ".dart": "dart",
    ".lua": "lua",
    ".pl": "perl", ".pm": "perl",
    ".r": "r",
    ".json": "json",
    ".md": "markdown", ".mdx": "markdown",
    ".html": "html", ".htm": "html",
    ".css": "css", ".scss": "scss", ".less": "less",
    ".xml": "xml", ".svg": "xml",
    ".yaml": "yaml", ".yml": "yaml",
    ".toml": "toml",
    ".sql": "sql",
    ".sh": "shell", ".bash": "shell", ".zsh": "shell",
    ".bat": "bat", ".cmd": "bat",
    ".ps1": "powershell",
    ".dockerfile": "dockerfile", "dockerfile": "dockerfile",
    ".gitignore": "ignore",
    ".env": "plaintext",
    ".txt": "plaintext",
    ".csv": "plaintext",
    ".log": "plaintext",
    ".diff": "diff", ".patch": "diff",
}


@dataclass
class FileBuffer:
    path: str
    content: str
    language: str = "plaintext"
    is_dirty: bool = False
    cursor_line: int = 1
    cursor_column: int = 1


class EditorManager:
    def __init__(self, db: Database | None = None):
        self._db = db
        self._open_files: dict[str, FileBuffer] = {}

    def open_file(self, path: str) -> FileBuffer | None:
        if not os.path.isfile(path):
            return None
        try:
            with open(path, "r", encoding="utf-8") as f:
                content = f.read()
        except (UnicodeDecodeError, Exception):
            return None

        lang = self._detect_language(path)
        buf = FileBuffer(path=path, content=content, language=lang)

        if path in self._open_files:
            self._open_files[path].content = content
            self._open_files[path].is_dirty = False
            self._open_files[path].language = lang
        else:
            self._open_files[path] = buf

        return buf

    def save_file(self, path: str, content: str) -> bool:
        try:
            with open(path, "w", encoding="utf-8") as f:
                f.write(content)
        except Exception:
            return False

        if path in self._open_files:
            self._open_files[path].content = content
            self._open_files[path].is_dirty = False
        else:
            lang = self._detect_language(path)
            self._open_files[path] = FileBuffer(path=path, content=content, language=lang)

        return True

    def close_file(self, path: str) -> bool:
        return self._open_files.pop(path, None) is not None

    def get_open_files(self) -> list[FileBuffer]:
        return list(self._open_files.values())

    def get_file(self, path: str) -> FileBuffer | None:
        return self._open_files.get(path)

    def detect_language(self, path: str) -> str:
        return self._detect_language(path)

    def update_cursor(self, path: str, line: int, column: int) -> bool:
        buf = self._open_files.get(path)
        if not buf:
            return False
        buf.cursor_line = line
        buf.cursor_column = column
        return True

    def get_settings(self) -> dict[str, str]:
        if not self._db:
            return _default_settings()
        rows = self._db.fetchall(
            "SELECT key, value FROM editor_settings ORDER BY key"
        )
        settings = dict(_default_settings())
        for r in rows:
            settings[r["key"]] = r["value"]
        return settings

    def update_setting(self, key: str, value: str) -> bool:
        if not self._db:
            return False
        self._db.execute(
            "INSERT OR REPLACE INTO editor_settings (key, value) VALUES (?, ?)",
            (key, value),
        )
        return True

    def _detect_language(self, path: str) -> str:
        name = os.path.basename(path).lower()
        if name in _LANG_MAP:
            return _LANG_MAP[name]
        ext = os.path.splitext(path)[1].lower()
        return _LANG_MAP.get(ext, "plaintext")


def _default_settings() -> dict[str, str]:
    return {
        "fontSize": "14",
        "tabSize": "4",
        "wordWrap": "off",
        "theme": "vs-dark",
        "minimap": "true",
        "lineNumbers": "on",
        "autoSave": "true",
        "autoSaveDelay": "2000",
    }
